Add Autocast Support for FakeTensors / use fake device dispatch keys

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82593
* #82571
* #82337
* __->__ #82449

From PR:
```
Note: [Fake Tensor Dispatch Keys]
In order to model the behavior of device-specific autocast 
and autograd logic, we update the dispatch keys of FakeTensors
to reflect their fake device. This includes the BackendComponent
(DispatchKey::Meta -> DispatchKey::CUDA), and also the BackendComponent
related Autocast and Autograd keys. __torch__dispatch__ sits below
Autocast and Autograd, and is only invoked when we are at the
kernel for the BackendComponent. Then, we add Meta to the 
thread-local dispatch include set to hit the meta kernel 
instead of the kernel of the BackendComponent for the fake device.
```

Also adds the `conv1/2/3d.padding` operators to the Autocast rule set. Without that fix, the FakeTensor dtype would diverge. 

See: https://github.com/pytorch/pytorch/issues/81608

eellison committed:
Add Autocast Support for FakeTensors / use fake device dispatch keys

[ghstack-poisoned]

eellison committed:
Update on "Add Autocast Support for FakeTensors / use fake device dispatch keys"




[ghstack-poisoned]

eellison committed:
Update on "Add Autocast Support for FakeTensors / use fake device dispatch keys"


From PR:
```
Note: [Fake Tensor Dispatch Keys]
In order to model the behavior of device-specific autocast 
and autograd logic, we update the dispatch keys of FakeTensors
to reflect their fake device. This includes the BackendComponent
(DispatchKey::Meta -> DispatchKey::CUDA), and also the BackendComponent
related Autocast and Autograd keys. __torch__dispatch__ sits below
Autocast and Autograd, and is only invoked when we are at the
kernel for the BackendComponent. Then, we add Meta to the 
thread-local dispatch include set to hit the meta kernel 
instead of the kernel of the BackendComponent for the fake device.
```

Also adds the `conv1/2/3d.padding` operators to the Autocast rule set. Without that fix, the FakeTensor dtype would diverge.

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -184,6 +186,35 @@ TensorImpl::TensorImpl(
+  // Keys]
+  key_set = key_set.remove_backend(old_backend);
+  key_set_ = key_set | DispatchKeySet(device_key);
+}
```
@bdhirsh can you review this part?

ezyang replied:
In particular, the getAutogradRelatedKeySetFromBackend feels due for a refactor; shouldn't this be as simple as just writing out the BC bit in the key set? Not sure why there is so much work that has to be done here.

bdhirsh replied:
I think you're right that all of the `getAutogradRelatedKeySetFromBackend()` calls above shouldn't be needed. @eellison did you run into issues before adding them?

bdhirsh replied:
Well... technically, NestedTensor has its own autograd key. I don't think that in practice any of the autograd kernels will impact the result of running meta tensor shape inference though, so maybe... we can just simplify this and not worry about the `AutogradNestedTensor` key here?

eellison replied:
For my tests you're right I can remove the `getAutogradRelatedKeySetFromBackend` related key logic

ezyang commented on torch/csrc/Module.cpp:
```diff
@@ -1258,6 +1259,29 @@ Call this whenever a new thread is created in order to propagate values from
+    auto k = key_set.highestBackendKey();
+    local_keyset.included_ = local_keyset.included_.remove_backend(k);
+    c10::impl::_force_tls_local_dispatch_key_set(local_keyset);
+  });
```
feels like it would be better to bind one of the local TLS RAII objects. It's pretty quick and easy

bdhirsh replied:
I'm trying to fix the functionalization <> dynamic shape issues and I ended up adding something pretty similar here: [_change_backend_component_keys.](https://github.com/pytorch/pytorch/pull/82453/files#diff-b3651b15177d065d4a02b0bd03703b6df569fc7a53ce88c4c6dbd6145adf35f6R89). If this looks good to you, then we can either stamp those changes into this PR or fix this code up after my PR lands.

eellison replied:
If you don't mind fixing up this code after my PR lands, that would be great. Thanks brian.

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -331,6 +334,19 @@ def __init__(self, fake_mode, elem, device: Union[torch.device, str]):
+        # thread-local dispatch include set to hit the meta kernel
+        # instead of the kernel of the BackendComponent for the fake device.
+
+        torch._C._change_backend_component_keys(self, device)
```
Yeah, my preference is probably to have this be an argument to `_make_subclass`, but if you need to land sooner than later this is OK. The big problem with this API is it lets people transmute backend key mid-Tensor lifetime and that's going to be very confusing.

eellison replied:
If you don't have access to expensive GPUs you can now `_change_backend_component_keys` on your CPU tensors to get access to all those great kernels 😉

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -358,17 +374,18 @@ def new(self, *args, **kwargs):
         # the call here
+        # because it doesn't go through the dispatcher, we run into errors
+        # when attempting to compute an output in meta, so
+        # we compute the real tensor then convert to meta
```
But this means that `new` is no longer memory efficient, is that right?

eellison replied:
Yea, this is sort of a deprecated API / doesn't go through the dispatcher, so it would make sense that it might be fragile to changes. Similar to fallbacks, seemed okay to allocate a real tensor temporarily.

ezyang replied:
OK. It's just odd that although we have explicit handling for it, we aren't able to just go ahead and write the "correct" fake rule for it. But I suppose it might be annoying to do.

bdhirsh commented on c10/core/TensorImpl.cpp:
```diff
@@ -184,6 +186,35 @@ TensorImpl::TensorImpl(
 
+void TensorImpl::_change_backend_component_keys(c10::Device device) {
+  DispatchKey device_key =
+      computeDispatchKey(c10::nullopt, c10::nullopt, device);
```
Feels like we should have a way to go directly from `Device` to `BackendComponent` - this feels a bit roundabout.

ezyang commented on aten/src/ATen/autocast_mode.cpp:
```diff
@@ -483,11 +483,15 @@ TORCH_LIBRARY_IMPL(_, AutocastCPU, m) {
   KERNEL_CPU(ADD_NS(conv2d), "conv2d", Tensor (const Tensor &, const Tensor &, const c10::optional<Tensor>&, IntArrayRef, IntArrayRef, IntArrayRef, int64_t), lower_precision_fp)
+  KERNEL_CPU(ADD_NS(conv2d), "conv2d.padding", Tensor (const Tensor&, const Tensor&, const c10::optional<Tensor>&, IntArrayRef, c10::string_view, IntArrayRef, int64_t groups), lower_precision_fp)
   KERNEL_CPU(ADD_NS(conv3d), "conv3d", Tensor (const Tensor &, const Tensor &, const c10::optional<Tensor>&, IntArrayRef, IntArrayRef, IntArrayRef, int64_t), lower_precision_fp)
+  KERNEL_CPU(ADD_NS(conv3d), "conv3d.padding", Tensor (const Tensor&, const Tensor&, const c10::optional<Tensor>&, IntArrayRef, c10::string_view, IntArrayRef, int64_t groups), lower_precision_fp)
```
Do you have a more detailed explanation about these? Is a bit confusing.

eellison replied:
`aten::conv2d.padding` redispatches to `aten::convolution`, which doesn't have an autocast registered for it, `_convolution` does. Its only in the CPU kernel by invoking `_convolution` that autocast gets applied. Since the meta kernel for `aten::convolution` wont call `_convolution` without the changes no dispatching gets applied.

So I mirrored the existing pattern for `conv1d` and added it to the registration list.

ezyang replied:
oof. I guess if we turned off autocast after we got past the autocast layer this would prevent that, but we don't really want to pay for it. :/ There may be other latent bugs like this.

bdhirsh replied:
At the risk of micro-optimizing: One thing we could maybe do if we’re worried about the boxing/unboxing cost of the fallback (although we’d still have to pay for the redispatch + TLS): since we don’t actually care about the operator’s arguments, we could teach the dispatcher to not box/unbox anything if the boxed fallback that you register doesn’t take in an argument stack (indicating that it doesn’t use the arguments)

bdhirsh commented on c10/core/TensorImpl.cpp:
```diff
@@ -184,6 +186,35 @@ TensorImpl::TensorImpl(
+
+  auto key_set =
+      key_set_ - c10::getAutocastRelatedKeySetFromBackend(old_backend);
+  key_set = key_set | c10::getAutocastRelatedKeySetFromBackend(new_backend);
```
I think these lines for autocast are necessary, but maybe add a TODO pointing out that Autocast should be a (per-backend) functionality key, and once we make that change then they'll no longer be necessary?

eellison committed:
Update on "Add Autocast Support for FakeTensors / use fake device dispatch keys"


From PR:
```
Note: [Fake Tensor Dispatch Keys]
In order to model the behavior of device-specific autocast 
and autograd logic, we update the dispatch keys of FakeTensors
to reflect their fake device. This includes the BackendComponent
(DispatchKey::Meta -> DispatchKey::CUDA), and also the BackendComponent
related Autocast and Autograd keys. __torch__dispatch__ sits below
Autocast and Autograd, and is only invoked when we are at the
kernel for the BackendComponent. Then, we add Meta to the 
thread-local dispatch include set to hit the meta kernel 
instead of the kernel of the BackendComponent for the fake device.
```

Also adds the `conv1/2/3d.padding` operators to the Autocast rule set. Without that fix, the FakeTensor dtype would diverge. 

See: https://github.com/pytorch/pytorch/issues/81608


[ghstack-poisoned]

eellison committed:
Update on "Add Autocast Support for FakeTensors / use fake device dispatch keys"


From PR:
```
Note: [Fake Tensor Dispatch Keys]
In order to model the behavior of device-specific autocast 
and autograd logic, we update the dispatch keys of FakeTensors
to reflect their fake device. This includes the BackendComponent
(DispatchKey::Meta -> DispatchKey::CUDA), and also the BackendComponent
related Autocast and Autograd keys. __torch__dispatch__ sits below
Autocast and Autograd, and is only invoked when we are at the
kernel for the BackendComponent. Then, we add Meta to the 
thread-local dispatch include set to hit the meta kernel 
instead of the kernel of the BackendComponent for the fake device.
```

Also adds the `conv1/2/3d.padding` operators to the Autocast rule set. Without that fix, the FakeTensor dtype would diverge. 

See: https://github.com/pytorch/pytorch/issues/81608


[ghstack-poisoned]

eellison committed:
Update on "Add Autocast Support for FakeTensors / use fake device dispatch keys"


From PR:
```
Note: [Fake Tensor Dispatch Keys]
In order to model the behavior of device-specific autocast 
and autograd logic, we update the dispatch keys of FakeTensors
to reflect their fake device. This includes the BackendComponent
(DispatchKey::Meta -> DispatchKey::CUDA), and also the BackendComponent
related Autocast and Autograd keys. __torch__dispatch__ sits below
Autocast and Autograd, and is only invoked when we are at the
kernel for the BackendComponent. Then, we add Meta to the 
thread-local dispatch include set to hit the meta kernel 
instead of the kernel of the BackendComponent for the fake device.
```

Also adds the `conv1/2/3d.padding` operators to the Autocast rule set. Without that fix, the FakeTensor dtype would diverge. 

See: https://github.com/pytorch/pytorch/issues/81608


[ghstack-poisoned]

eellison committed:
Update on "Add Autocast Support for FakeTensors / use fake device dispatch keys"


From PR:
```
Note: [Fake Tensor Dispatch Keys]
In order to model the behavior of device-specific autocast 
and autograd logic, we update the dispatch keys of FakeTensors
to reflect their fake device. This includes the BackendComponent
(DispatchKey::Meta -> DispatchKey::CUDA), and also the BackendComponent
related Autocast and Autograd keys. __torch__dispatch__ sits below
Autocast and Autograd, and is only invoked when we are at the
kernel for the BackendComponent. Then, we add Meta to the 
thread-local dispatch include set to hit the meta kernel 
instead of the kernel of the BackendComponent for the fake device.
```

Also adds the `conv1/2/3d.padding` operators to the Autocast rule set. Without that fix, the FakeTensor dtype would diverge. 

See: https://github.com/pytorch/pytorch/issues/81608


[ghstack-poisoned]

eellison committed:
Update on "Add Autocast Support for FakeTensors / use fake device dispatch keys"


From PR:
```
Note: [Fake Tensor Dispatch Keys]
In order to model the behavior of device-specific autocast 
and autograd logic, we update the dispatch keys of FakeTensors
to reflect their fake device. This includes the BackendComponent
(DispatchKey::Meta -> DispatchKey::CUDA), and also the BackendComponent
related Autocast and Autograd keys. __torch__dispatch__ sits below
Autocast and Autograd, and is only invoked when we are at the
kernel for the BackendComponent. Then, we add Meta to the 
thread-local dispatch include set to hit the meta kernel 
instead of the kernel of the BackendComponent for the fake device.
```

Also adds the `conv1/2/3d.padding` operators to the Autocast rule set. Without that fix, the FakeTensor dtype would diverge. 

See: https://github.com/pytorch/pytorch/issues/81608


[ghstack-poisoned]

eellison commented:
@pytorchbot merge -g
